File: epi_judge_python/search_first_key.py
from test_framework import generic_test


def search_first_of_k(A, k):
    l, r = 0, len(A) - 1

    curr = -1
    while l <= r:
        m = (l + r)//2
        if A[m] == k:
            curr = m
            r = m - 1
        elif A[m] < k:
            l = m + 1
        else:
            r = m - 1

    return curr


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            "search_first_key.py", 'search_first_key.tsv', search_first_of_k))

File: epi_judge_java/epi/SearchFirstKey.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
import java.util.List;
public class SearchFirstKey {
  @EpiTest(testDataFile = "search_first_key.tsv")

  public static int searchFirstOfK(List<Integer> A, int k) {
    int l,r, candidate;
    l = 0;
    r = A.size() - 1;
    candidate = -1;

    while (l <= r){
      int m = (l +r)/2;
      if (A.get(m) == k){
        candidate = m;
        r = m - 1;
      } else if (A.get(m) < k){
        l = m + 1;
      } else {
        r = m - 1;
      }
    }
    return candidate;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SearchFirstKey.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
